dvc/stage.py

```diff
@@ -603,6 +611,10 @@ def checkout(self, force=False):
         for out in self.outs:
             out.checkout(force=force)
 
+    def _already_cached(self):
```

Comment:
1) We should also check if md5 of the stage didn't change(see `stage.changed()`).
2) Checking if `out.cache` is not None or empty is also not enough since the cache might be bad(corrupted cache case). So we need to implement and use something like `out.cache_changed()` instead. All remote drivers have `changed_cache()` implemented already, so that should help.